From 96bd4737326f53d97473fc429721f1b70f5312db Mon Sep 17 00:00:00 2001
From: David McKay <david.mckay@codasip.com>
Date: Tue, 19 Mar 2024 13:23:39 +0000
Subject: [PATCH 08/15] riscv: hobgoblin: Xilinx EthernetLite support

This patch add in support for the Xilinx EthernetLite IP on the
GenesysII board. This is a very simple Ethernet controller. It does not
do DMA, so all data is transferred via the CPU. Performance is low,
especially as the CPU on the hobgoblin platform is only running
at 50MHz or so.

The dts file is called a70x-hobgoblin-etherlite.dts as there will be a
version using the Xilinx axi-ethernet which is a much more capable
controller, but requires a license to build. The non ethernet version is
removed from the build to remove clutter.

Signed-off-by: David McKay <david.mckay@codasip.com>
---
 arch/riscv/boot/dts/codasip/Makefile          |  2 +-
 .../dts/codasip/a70x-hobgoblin-etherlite.dts  | 36 +++++++++++++++++++
 .../configs/codasip-a70x-hobgoblin_defconfig  |  2 ++
 3 files changed, 39 insertions(+), 1 deletion(-)
 create mode 100644 arch/riscv/boot/dts/codasip/a70x-hobgoblin-etherlite.dts

diff --git a/arch/riscv/boot/dts/codasip/Makefile b/arch/riscv/boot/dts/codasip/Makefile
index 79c1066932c2..6df53f663347 100644
--- a/arch/riscv/boot/dts/codasip/Makefile
+++ b/arch/riscv/boot/dts/codasip/Makefile
@@ -1,3 +1,3 @@
 # SPDX-License-Identifier: GPL-2.0
-dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin.dtb
 dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin-qemu.dtb
+dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin-etherlite.dtb
diff --git a/arch/riscv/boot/dts/codasip/a70x-hobgoblin-etherlite.dts b/arch/riscv/boot/dts/codasip/a70x-hobgoblin-etherlite.dts
new file mode 100644
index 000000000000..1cbe38fe04e0
--- /dev/null
+++ b/arch/riscv/boot/dts/codasip/a70x-hobgoblin-etherlite.dts
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 Codasip Limited
+ *
+ * A730 Hobgoblin FPGA platform emulation in qemu.
+ */
+
+#include "a70x-hobgoblin.dts"
+
+&soc {
+	axi_ethernetlite: ethernet@60020000 {
+		compatible = "xlnx,xps-ethernetlite-3.00.a";
+		device_type = "network";
+		interrupts-extended = <&plic 15>;
+		phy-handle = <&phy0>;
+		reg = <0x60020000 0x2000>;
+		xlnx,duplex = <0x1>;
+		xlnx,include-global-buffers = <0x1>;
+		xlnx,include-internal-loopback = <0x1>;
+		xlnx,include-mdio = <0x1>;
+		xlnx,instance = "axi_ethernetlite_inst";
+		xlnx,rx-ping-pong = <0x1>;
+		xlnx,s-axi-id-width = <0x0>;
+		xlnx,tx-ping-pong = <0x1>;
+		xlnx,use-internal = <0x0>;
+
+		mdio {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			phy0: phy@1 {
+				device_type = "ethernet-phy";
+				reg = <1>;
+			      };
+		};
+	};
+};
diff --git a/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig b/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
index c7d97cbbf52e..7e2491e8fbe9 100644
--- a/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
+++ b/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
@@ -54,6 +54,8 @@ CONFIG_BLK_DEV_LOOP=y
 CONFIG_VIRTIO_BLK=y
 CONFIG_NETDEVICES=y
 CONFIG_VIRTIO_NET=y
+CONFIG_XILINX_EMACLITE=y
+CONFIG_REALTEK_PHY=y
 # CONFIG_WLAN is not set
 CONFIG_INPUT_MOUSEDEV=y
 CONFIG_KEYBOARD_GPIO=y
-- 
2.39.3

